### PR TITLE
Android: Access to files from APK is boosted

### PR DESCRIPTION
**Pull request by mingulov**
Open APK file just once and cache a file list information (name and position inside archive). Also a general ZipFile helper class is created to read multiple files from Zip archives - might be used for another platforms if needed.

Previously it was so for each separate file access operation:
- Open zip archive, initially parse its structure (unzOpen)
- Locate the particular file there by a linear search through full archive (unzLocateFile).
  So if file does not exist - still an archive has to be handled fully, even slower than a file exists.
- Read file.
- Close zip archive

After this commit:
- Once - open zip/apk file and collect file list information
  (average consumed time - the same like a search for 2 different files by unzLocateFile)
- When needed - directly retrieve file position inside the archive, setting it up to the zip reader and reading file

ZipFile class is located in support/zip_support/ZipUtils.h and .cpp to prevent creation of a new files and adding them to multiple project files.

std::map is used to store file list information, if supported by a platform - might be changed to std::unordered_map locally (defined at ZipUtils.cpp).

ZipFIle class declaration uses private implementation to prevent overloading header with a lot of other headers (unzip.h etc).

Speedup is visible even on just a simple TestCpp project on TextureCacheTest. On more complex projects with a lot assets / multiple loading / reloading operations - speedup is much more significant.

### CODE REVIEW

File: cocos2dx/support/zip_support/ZipUtils.cpp [hunk 1]
```diff
@@ -280,4 +284,139 @@ int ZipUtils::ccInflateCCZFile(const char *path, unsigned char **out)
...
+        CC_BREAK_IF(it ==  m_data->fileList.end());
+
+        unz_file_pos filePos = it->second;
+
+        int nRet = unzGoToFilePos(m_data->zipFile, &filePos);
+        CC_BREAK_IF(UNZ_OK != nRet);
+
+        char szFilePathA[UNZ_MAXFILENAMEINZIP + 1];
+        unz_file_info FileInfo;
+        nRet = unzGetCurrentFileInfo(m_data->zipFile, &FileInfo,
```
Comment (minggo) [inline cocos2dx/support/zip_support/ZipUtils.cpp]: It seems that szFilePathA is not needed, can we pass NULL to unzGetCurrentFileInfo()?

File: cocos2dx/support/zip_support/ZipUtils.cpp [hunk 2]
```diff
@@ -280,4 +284,139 @@ int ZipUtils::ccInflateCCZFile(const char *path, unsigned char **out)
...
+                szFilePathA, sizeof(szFilePathA) - 1, NULL, 0, NULL, 0);
+        CC_BREAK_IF(UNZ_OK != nRet);
+
+        nRet = unzOpenCurrentFile(m_data->zipFile);
+        CC_BREAK_IF(UNZ_OK != nRet);
+
+        pBuffer = new unsigned char[FileInfo.uncompressed_size];
```
Comment (minggo) [inline cocos2dx/support/zip_support/ZipUtils.cpp]: Can cache uncompressed_size too?

File: cocos2dx/platform/android/CCFileUtils.cpp [hunk 3]
```diff
@@ -43,6 +45,7 @@ CCFileUtils* CCFileUtils::sharedFileUtils()
     {
         s_pFileUtils = new CCFileUtils();
         s_strResourcePath = getApkPath();
+        s_pZipFile = new ZipFile(s_strResourcePath, "assets/");
```
Comment (minggo) [inline cocos2dx/platform/android/CCFileUtils.cpp]: s_pZipFile is not deleted?

**Comment by minggo** [comment]
It is an awesome work.
Thank you.

File: cocos2dx/support/zip_support/ZipUtils.cpp (hunk 2 above)
Comment (mingulov) [inline cocos2dx/support/zip_support/ZipUtils.cpp]: Thank you, a call to unzGetCurrentFileInfo at ZipFile::getFileData is removed, uncompressed_size is cached at ZipFile::setFilter together with file pos info.

File: cocos2dx/platform/android/CCFileUtils.cpp (hunk 3 above)
Comment (mingulov) [inline cocos2dx/platform/android/CCFileUtils.cpp]: Fixed, will be deleted at CCFileUtils::purgeFileUtils together with s_pFileUtils.

**Pull request merged**